index.js

```diff
@@ -43,89 +60,234 @@ function cmdShim (src, to, opts) {
     .then(() => cmdShim_(src, to, opts))
 }
 
+/**
+ * Try to create shims **even if `src` is missing**.
+ *
+ * @param {string} src Path to program (executable or script).
+ * @param {string} to Path to shims.
+ * Don't add an extension if you will create multiple types of shims.
+ * @param {Options} opts Options.
+ *
+ */
 function cmdShim_ (src, to, opts) {
-  return Promise.all([
-    rm(to),
-    rm(`${to}.ps1`),
-    opts.createCmdFile && rm(`${to}.cmd`)
-  ])
-    .then(() => writeShim(src, to, opts))
+  opts = Object.assign({}, DEFAULT_OPTIONS, opts)
+  return searchScriptRuntime(src).then((srcRuntimeInfo) => {
+    // Always tries to create all types of shims by calling `writeAllShims` as of now.
+    // Append your code here to change the behavior in response to `srcRuntimeInfo`.
+
+    // Create 3 shims for (Ba)sh in Cygwin / MSYS, no extension) & CMD (.cmd) & PowerShell (.ps1)
+    return writeShimsPreCommon(to).then(() => writeAllShims(src, to, srcRuntimeInfo, opts))
+  })
+}
+
+/**
+ * Do processes before **all** shims are created.
+ * This must be called **only once** for one call of `cmdShim(IfExists)`.
+ *
+ * @param {string} target Path of shims that are going to be created.
+ */
+function writeShimsPreCommon (target) {
+  return mkdir(path.dirname(target))
 }
 
-function writeShim (src, to, opts) {
+/**
+ * Write all types (sh & cmd & pwsh) of shims to files.
+ * Extensions (`.cmd` and `.ps1`) are appended to cmd and pwsh shims.
+ *
+ *
+ * @param {string} src Path to program (executable or script).
+ * @param {string} to Path to shims **without extensions**.
+ * Extensions are added for CMD and PowerShell shims.
+ * @param {RuntimeInfo} srcRuntimeInfo Return value of `await searchScriptRuntime(src)`.
+ * @param {Options} opts Options.
+ */
+function writeAllShims (src, to, srcRuntimeInfo, opts) {
   opts = Object.assign({}, DEFAULT_OPTIONS, opts)
-  const defaultArgs = opts.preserveSymlinks ? '--preserve-symlinks' : ''
-  // make a cmd file and a sh script
-  // First, check if the bin is a #! of some sort.
-  // If not, then assume it's something that'll be compiled, or some other
-  // sort of script, and just call it directly.
-  return mkdir(path.dirname(to))
-    .then(() => {
-      return fs.readFile(src, 'utf8')
-        .then(data => {
-          const firstLine = data.trim().split(/\r*\n/)[0]
-          const shebang = firstLine.match(shebangExpr)
-          if (!shebang) return writeShim_(src, to, Object.assign({}, opts, {args: defaultArgs}))
-          const prog = shebang[1]
-          const args = (shebang[2] && ((defaultArgs && (shebang[2] + ' ' + defaultArgs)) || shebang[2])) || defaultArgs
-          return writeShim_(src, to, Object.assign({}, opts, {prog, args}))
+  const generatorAndExtPairs = [[generateShShim, '']]
```

Comment:
```suggestion
  /** @type {Array<[ShimGenerator, string]>} */
  const generatorAndExtPairs = [[generateShShim, '']]
```
Without this, the type ends up being `(string|function)[][]` rather than an&nbsp;array of&nbsp;fixed size arrays where the&nbsp;first element is&nbsp;of&nbsp;type `ShimGenerator` and the&nbsp;second is&nbsp;of&nbsp;type&nbsp;`string`.